### DISCUSSION TITLE
Exposing textDocument/highlight as a function to go on eldoc-documentation-functions

### DISCUSSION THREAD

**Discussion by bigodel** [joaotavora/eglot#1398] [Ideas]
As the title says, I'd like to have the function that highlights symbols in the document (currently that'd be `eglot--highlight-piggyback`) be public to be used in `eldoc-documentation-functions`. The reasoning for it is that, ideally, `eldoc-documentation-functions` would have only `eglot-eldoc-signature-function` to show the function's signature (although I know many language servers have a very poor implementation of `textDocument/signatureHelp` so I do understand why `eglot-eldoc-hover-function` is there by default), so it would be interesting to enable highlighting without having to enable hover. `eldoc-documentation-functions` might not be the best place for this, though, but I can't think of any other place to put it (`post-command-hook` maybe?).

**Comment by joaotavora** [comment]
>  so it would be interesting to enable highlighting without having to enable hover. eldoc-documentation-functions might not be the best place for this, though, but I can't think of any other place to put it (post-command-hook maybe?).

I think you need to provide, without reference to the functions or internals, an example where you describe exactly what happens *today* (in the server of your choice, with an example file of your choice) and what happens *tomorrow* when your dream change is implemented.  I found it hard to follow your text, though I recognize some parts.

**Reply by bigodel** [reply]
I'll try to rephrase my proposal: Instead of having `eglot--highlight-piggyback` **inside** `eglot-hover-eldoc-function`, the proposal is to instead rewrite it as `eglot-highlight-eldoc-function` and place it inside `eldoc-documentation-functions`, so its default value on an Eglot managed buffer would be

```elisp
'(eglot-signature-eldoc-function
  eglot-highlight-eldoc-function
  elgot-hover-eldoc-function)
```

I, for one, manage `eldoc-documentation-functions` myself to remove `eglot-hover-eldoc-function`, but I'd still like to have highlighting of the symbol at point. I managed to achieve that by having the variable's value as

```elisp
'(eglot-signature-eldoc-function
  eglot--highlight-piggyback)
```

Of course putting the highlighting function in `eldoc-documentation-functions` isn't ideal, as evidenced by the FIXME comment at the top of the definition of `eglot--highlight-piggyback`, but I don't have any idea on where to put it besides the ElDoc variable.

I'll try to provide an example in the next comment.

**Reply by bigodel** [reply]
Here's the current behaviour, notice how the symbol `baz` is not highlighted because `elgot-hover-eldoc-function` is not present in `eldoc-documentation-functions`:

![image](https://github.com/joaotavora/eglot/assets/23617959/264c2b35-158f-4aba-b4c6-9693557ab5b8)

And here is the desired behaviour:

![image](https://github.com/joaotavora/eglot/assets/23617959/37cc99d6-03c4-4e81-948d-7999e016aa5a)

where even though `eglot-hover-eldoc-function` is not part of `eldoc-documentation-functions` I can still achieve highlighting of the symbol at point.

I used `clojure-lsp` as the server, but that would be applicable to any server.

**Reply by joaotavora** [reply]
OK, I see the use case and this could work.  But not, this isn't the right way.  the piggyback thing is a hack and I prefer to keep it a known hack instead of enshrining it for the future.  You could define 

```lisp
(defun bigodel/eldoc-thingy (cb)
   (eglot-hover-eldoc-function cb)
   nil)
```

And then add _that_ to `e-d-f`.  If you don't want to pay the cost of requesting an unneeded hover, you can propose probably this patch to the bug reporting list:

```diff
diff --git a/lisp/progmodes/eglot.el b/lisp/progmodes/eglot.el
index 6896baf30ce..6470232c6d5 100644
--- a/lisp/progmodes/eglot.el
+++ b/lisp/progmodes/eglot.el
@@ -3482,6 +3482,7 @@ eglot-signature-eldoc-function
 
 (defun eglot-hover-eldoc-function (cb)
   "A member of `eldoc-documentation-functions', for hover."
+  (eglot--highlight-piggyback cb)
   (when (eglot-server-capable :hoverProvider)
     (let ((buf (current-buffer)))
       (jsonrpc-async-request
@@ -3494,7 +3495,6 @@ eglot-hover-eldoc-function
                          (funcall cb info
                                   :echo (and info (string-match "\n" info))))))
        :deferred :textDocument/hover))
-    (eglot--highlight-piggyback cb)
     t))
 
 (defvar eglot--highlights nil "Overlays for textDocument/documentHighlight.")
```

and then add `:hoverProvider` to `eglot-ignored-server-capabilities`.

**Reply by bigodel** [reply]
> OK, I see the use case and this could work. But not, this isn't the right way. the piggyback thing is a hack and I prefer to keep it a known hack instead of enshrining it for the future. 

I agree with you on this one, but I think we could expose the highlighting functionality in a way that's decoupled from `eglot-hover-eldoc-function`, either in a `post-command-hook` or inside the body of `eglot--change-idle-timer` (this is probably the best place, having read a bit more of the source code). It could be disabled by just adding `:documentHighlightProvider` to `eglot-ignored-server-capabilities`, without interfering with `:hoverProvider`.

> You could define
> 
> ```commonlisp
> (defun bigodel/eldoc-thingy (cb)
>    (eglot-hover-eldoc-function cb)
>    nil)
> ```
> 
> And then add _that_ to `e-d-f`. If you don't want to pay the cost of requesting an unneeded hover, you can propose probably this patch to the bug reporting list:
> 
> ```diff
> diff --git a/lisp/progmodes/eglot.el b/lisp/progmodes/eglot.el
> index 6896baf30ce..6470232c6d5 100644
> --- a/lisp/progmodes/eglot.el
> +++ b/lisp/progmodes/eglot.el
> @@ -3482,6 +3482,7 @@ eglot-signature-eldoc-function
>  
>  (defun eglot-hover-eldoc-function (cb)
>    "A member of `eldoc-documentation-functions', for hover."
> +  (eglot--highlight-piggyback cb)
>    (when (eglot-server-capable :hoverProvider)
>      (let ((buf (current-buffer)))
>        (jsonrpc-async-request
> @@ -3494,7 +3495,6 @@ eglot-hover-eldoc-function
>                           (funcall cb info
>                                    :echo (and info (string-match "\n" info))))))
>         :deferred :textDocument/hover))
> -    (eglot--highlight-piggyback cb)
>      t))
>  
>  (defvar eglot--highlights nil "Overlays for textDocument/documentHighlight.")
> ```
> 
> and then add `:hoverProvider` to `eglot-ignored-server-capabilities`.

But I don't want to disable `:hoverProvider`, I just don't want its information to be displayed in the echo area. I'm actually using a combination of your suggested [persistent ElDoc buffer command](https://github.com/joaotavora/eglot/discussions/931#discussioncomment-2628625) with [this hack](https://github.com/casouri/eldoc-box/issues/90#issuecomment-1751968271) to use `eglot-hover-eldoc-function` only when displaying the ElDoc buffer (I'm working on an implementation of that hack using `eldoc-documentation-strategy`, but the hack works for now).

**Reply by bigodel** [reply]
Any thoughts on how to implement `:documentHighlightProvider` independent from `:hoverProvider`? It could likely be implemented in a way that could still be reused by `eglot-hover-eldoc-function`.

BTW why they are couple together? Seems odd given that they are two different `textDocument` capabilities.

**Reply by joaotavora** [reply]
They are coupled together to save on UI logic, hence the name "piggyback".

As far as I understand, your desired behavior as described earlier can be solved by `bigodel/eldoc-thingy` as suggested.

**Reply by bigodel** [reply]
> As far as I understand, your desired behavior as described earlier can be solved by `bigodel/eldoc-thingy` as suggested.

Nope, it doesn't, since doing that will require a full hover when I only want `:documentHighlightProvider`. And your other proposal also doesn't work because I don't want to fully disable `:hoverProvider`, I only disable it when ElDoc is being presented in the echo area. I am requesting and willing to propose a patch that exposes `:documentHighlightProvider` as a feature for users independently to `:hoverProvider`.

**Reply by joaotavora** [reply]
You told me 

>  even though eglot-hover-eldoc-function is not part of eldoc-documentation-functions I can still achieve highlighting of the symbol at point.

replace eglot-hover-eldoc-function with the functtion I gave you and you will have this.  This is all the help I can give you.

**Reply by bigodel** [reply]
> replace eglot-hover-eldoc-function with the functtion I gave you and you will have this.

I tried it, and that doesn't work consistently as some times (I can't pinpoint exactly when it happens) it'll bring up hover information to the echo area, which I'm trying to avoid. Besides, it triggers a full hover when all I want is to highlight the symbol at point...

>  This is all the help I can give you.

Can I take this as you don't think it would be beneficial to provide `:documentHighlightProvider` as a feature? Is there any particular reasoning besides saving a little bit of UI logic?

**Comment by bigodel** [comment]
This was introduced as `eglot-highlight-eldoc-function`.
